plugins/askrene/askrene.c

```diff
@@ -532,61 +472,164 @@ void get_constraints(const struct route_query *rq,
 	reserve_sub(rq->reserved, &scidd, max);
 }
 
-struct getroutes_info {
-	struct command *cmd;
-	struct node_id *source, *dest;
-	struct amount_msat *amount, *maxfee;
-	u32 *finalcltv, *maxdelay;
-	const char **layers;
-	struct additional_cost_htable *additional_costs;
-	/* Non-NULL if we are told to use "auto.localchans" */
-	struct layer *local_layer;
-};
+static void json_add_getroutes(
+    struct json_stream *js,
+    struct route **routes,
+    /* FIXME: can't we deduce these amounts from the routes themselves */
+    struct amount_msat *amounts,
+    double probability,
+    /* FIXME: can't we have different final_cltv for every route? */
+    u32 final_cltv)
+{
+	json_add_u64(js, "probability_ppm", (u64)(probability * 1000000));
+	json_array_start(js, "routes");
+	for (size_t i = 0; i < tal_count(routes); i++) {
+		json_object_start(js, NULL);
+		json_add_u64(js, "probability_ppm",
+			     (u64)(routes[i]->success_prob * 1000000));
+		json_add_amount_msat(js, "amount_msat", amounts[i]);
+		json_add_u32(js, "final_cltv", final_cltv);
+		json_array_start(js, "path");
+		for (size_t j = 0; j < tal_count(routes[i]->hops); j++) {
+			struct short_channel_id_dir scidd;
+			const struct route_hop *r = &routes[i]->hops[j];
+			json_object_start(js, NULL);
+			scidd.scid = r->scid;
+			scidd.dir = r->direction;
+			json_add_short_channel_id_dir(
+			    js, "short_channel_id_dir", scidd);
+			json_add_node_id(js, "next_node_id", &r->node_id);
+			json_add_amount_msat(js, "amount_msat", r->amount);
+			json_add_u32(js, "delay", r->delay);
+			json_object_end(js);
+		}
+		json_array_end(js);
+		json_object_end(js);
+	}
+	json_array_end(js);
+}
 
 static struct command_result *do_getroutes(struct command *cmd,
 					   struct gossmap_localmods *localmods,
 					   const struct getroutes_info *info)
 {
+	bool localmods_applied = false;
```

Comment:
Static flows are always preferred over dynamic flows.  They allow the reader, and the compiler, to do simple analysis which includes spotting uninitialized variables.

Reply:
In this case I think it's always set in the error path anyway?

Reply:
Yes. In this case it is trivial.
I was just wary that someone in the future would add another `goto fail` before `gossmap_apply_localmods`
leading to a call to `gossmap_remove_localmods` when `apply` was never called.

I am fixing this according to your comment.